include/aws/crt/Variant.h

```diff
@@ -0,0 +1,595 @@
+                    "This variant does not have such alternative T.");
+                static_assert(
+                    sizeof(T) <= STORAGE_SIZE,
+                    "Attempting to instantiate a Variant with a type bigger than all alternatives.");
+
+                using PlainT = typename std::decay<T>::type;
+                new (m_storage) PlainT(std::forward<Args>(args)...);
+                m_index = VariantDetail::Index::GetIndexOf<T, Ts...>();
+                AWS_ASSERT(m_index != -1);
+            }
+
+            Variant &operator=(const Variant &other)
+            {
+                if (this != &other)
+                {
+                    if (m_index != other.m_index)
+                    {
+                        Destroy();
+                        CopyConstructUtil<0, Ts...>(other);
+                    }
+                    else
+                    {
+                        CopyAssignUtil<0, Ts...>(other);
+                    }
+                }
+                return *this;
+            }
+
+            Variant &operator=(Variant &&other)
+            {
+                if (this != &other)
+                {
+                    if (m_index != other.m_index)
+                    {
+                        Destroy();
+                        MoveConstructUtil<0, Ts...>(std::move(other));
+                    }
+                    else
+                    {
+                        MoveAssignUtil<0, Ts...>(std::move(other));
+                    };
+                }
+                return *this;
+            }
+
+            /* emplace */
+            template <typename T, typename... Args, EnableIfOtherIsThisVariantAlternative<T> = 1>
+            T &emplace(Args &&...args)
+            {
+                static_assert(
+                    VariantDetail::Checker::HasType<typename std::decay<T>::type, Ts...>::value,
+                    "This variant does not have such alternative T.");
+                static_assert(
+                    sizeof(T) <= STORAGE_SIZE,
+                    "Attempting to instantiate a Variant with a type bigger than all alternatives.");
+
+                Destroy();
+
+                using PlainT = typename std::decay<T>::type;
+                new (m_storage) PlainT(std::forward<Args>(args)...);
+                m_index = VariantDetail::Index::GetIndexOf<T, Ts...>();
+                AWS_ASSERT(m_index != -1);
+
+                T *value = reinterpret_cast<T *>(m_storage);
+                return *value;
+            }
+
+            template <std::size_t Index, typename... Args>
+            auto emplace(Args &&...args) -> typename ThisVariantAlternative<Index>::type &
+            {
+                static_assert(Index < AlternativeCount, "Unknown alternative index to emplace");
+                using AlternativeT = typename ThisVariantAlternative<Index>::type;
+
+                return emplace<AlternativeT, Args...>(args...);
```

Comment:
`args` should be forwarded here, i.e. `std::forward<Args>(args)...`.
Also, there are no tests for this version of `emplace`.

Reply:
Good catch, thank you